lib/archethic/utxo.ex

```diff
@@ -0,0 +1,124 @@
+  def load_transaction(tx = %Transaction{}) do
+    authorized_nodes = P2P.authorized_and_available_nodes()
+
+    # Ingest all the movements to fill up the UTXO list
+    ingest_utxo(tx, authorized_nodes)
+
+    # Consume the transaction to update the unspent outputs from the consumed inputs
+    consume_utxo(tx, authorized_nodes)
+
+    Logger.info("Loaded into in memory UTXO tables",
+      transaction_address: Base.encode16(tx.address),
+      transaction_type: tx.type
+    )
+  end
+
+  defp ingest_utxo(
+         %Transaction{
+           address: address,
+           type: tx_type,
+           validation_stamp: %ValidationStamp{
+             protocol_version: protocol_version,
+             timestamp: timestamp,
+             ledger_operations: %LedgerOperations{transaction_movements: transaction_movements}
+           }
+         },
+         authorized_nodes
+       ) do
+    transaction_movements
+    |> Enum.map(&TransactionMovement.maybe_convert_reward_to_uco(&1, tx_type))
+    # |> TransactionMovements.aggregate() after PR about aggregation
+    |> Enum.each(fn %TransactionMovement{to: to, amount: amount, type: type} ->
+      genesis_address = DB.get_genesis_address(to)
+
+      if genesis_node?(genesis_address, authorized_nodes) do
+        utxo = %VersionedUnspentOutput{
+          unspent_output: %UnspentOutput{
+            from: address,
+            amount: amount,
+            timestamp: timestamp,
+            type: type
+          },
+          protocol_version: protocol_version
+        }
+
+        Loader.add_utxo(utxo, genesis_address)
+      end
+    end)
+  end
+
+  defp consume_utxo(tx = %Transaction{}, authorized_nodes) do
+    case find_genesis_address(tx) do
+      {:ok, genesis_address} ->
+        if genesis_node?(genesis_address, authorized_nodes) do
+          Loader.consume_inputs(tx, genesis_address)
+        end
+
+      _ ->
+        # ignore if genesis's address is not found
+        :ok
+    end
+  end
+
+  defp find_genesis_address(tx = %Transaction{address: address}) do
+    case DB.find_genesis_address(address) do
+      {:ok, genesis_address} ->
+        # This happens when the last transaction is ingested in the system (i.e last's tx chain)
+        {:ok, genesis_address}
+
+      {:error, :not_found} ->
+        # This might happens when the transaction haven't been yet synchronized but the previous transaction is already in the system (i.e genesis's chain)
+        tx
+        |> Transaction.previous_address()
+        |> DB.find_genesis_address()
+    end
+  end
+
+  defp genesis_node?(genesis_address, nodes) do
+    genesis_nodes = Election.chain_storage_nodes(genesis_address, nodes)
+    Utils.key_in_node_list?(genesis_nodes, Crypto.first_node_public_key())
+  end
+
+  @doc """
+  Returns the list of all the inputs which have not been consumed for the given chain's address
+  """
+  @spec get_unspent_outputs(binary()) :: list(VersionedUnspentOutput.t())
+  def get_unspent_outputs(address) do
```

Comment:
I propose to add typespec: `Enumerable.t() | list(VersionedUnspentOutput.t())`, because at the end List or Stream are enumerable

For the stream, I can add it, I think it might be useful for serialization or limitation through request streaming, because as the end it will only contain 100 inputs per genesis in memory